cleanrl/ppo_atari_envpool_xla_jax.py

```diff
@@ -0,0 +1,447 @@
+        b_actions = storage.actions.reshape((-1,) + envs.single_action_space.shape)
+        b_advantages = storage.advantages.reshape(-1)
+        b_returns = storage.returns.reshape(-1)
+
+        def ppo_loss(params, x, a, logp, mb_advantages, mb_returns):
+            newlogprob, entropy, newvalue = get_action_and_value2(params, x, a)
+            logratio = newlogprob - logp
+            ratio = jnp.exp(logratio)
+            approx_kl = ((ratio - 1) - logratio).mean()
+
+            if args.norm_adv:
+                mb_advantages = (mb_advantages - mb_advantages.mean()) / (mb_advantages.std() + 1e-8)
+
+            # Policy loss
+            pg_loss1 = -mb_advantages * ratio
+            pg_loss2 = -mb_advantages * jnp.clip(ratio, 1 - args.clip_coef, 1 + args.clip_coef)
+            pg_loss = jnp.maximum(pg_loss1, pg_loss2).mean()
+
+            # Value loss
+            v_loss = 0.5 * ((newvalue - mb_returns) ** 2).mean()
+
+            entropy_loss = entropy.mean()
+            loss = pg_loss - args.ent_coef * entropy_loss + v_loss * args.vf_coef
+            return loss, (pg_loss, v_loss, entropy_loss, jax.lax.stop_gradient(approx_kl))
+
+        ppo_loss_grad_fn = jax.value_and_grad(ppo_loss, has_aux=True)
+
+        # clipfracs = []
+        for _ in range(args.update_epochs):
+            key, subkey = jax.random.split(key)
+            b_inds = jax.random.permutation(subkey, args.batch_size, independent=True)
+            for start in range(0, args.batch_size, args.minibatch_size):
+                end = start + args.minibatch_size
+                mb_inds = b_inds[start:end]
+                (loss, (pg_loss, v_loss, entropy_loss, approx_kl)), grads = ppo_loss_grad_fn(
+                    agent_state.params,
+                    b_obs[mb_inds],
+                    b_actions[mb_inds],
+                    b_logprobs[mb_inds],
+                    b_advantages[mb_inds],
+                    b_returns[mb_inds],
+                )
+                agent_state = agent_state.apply_gradients(grads=grads)
+        return agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    start_time = time.time()
+    next_obs = envs.reset()
+    next_done = np.zeros(args.num_envs)
+
+    @jax.jit
+    def rollout(agent_state, episode_stats, next_obs, next_done, storage, key, handle, global_step):
+        for step in range(0, args.num_steps):
+            global_step += 1 * args.num_envs
+            storage, action, key = get_action_and_value(agent_state, next_obs, next_done, storage, step, key)
+
+            # TRY NOT TO MODIFY: execute the game and log data.
+            episode_stats, handle, (next_obs, reward, next_done, _) = step_env_wrappeed(episode_stats, handle, action)
+            storage = storage.replace(rewards=storage.rewards.at[step].set(reward))
+        return agent_state, episode_stats, next_obs, next_done, storage, key, handle, global_step
```

Comment:
Yeah this is a tricky situation. If we don't do this, then the rollout function eneds to accept `args.num_steps` as part of the args, because `args` itself is not a numpy array that is jittable. I did this to sort of minimize the difference between the `ppo_atari_envpool.py`, but perhaps it warrents a refactor here.

Reply:
More refactoring is coming, so let's spare doing it in this PR.